library(data.table)

# Read in file with the number of reads per transcript
dat <- read.csv("output/ENST_num_reads.txt")

# Indices for the SLC25 family transcripts
# Double check the mapping between the RefSeq IDs and Ensembl IDs are accurate
# mRNA sequences are hosted on NCBI, but index uses ensembl annotation
SLC25 <- c("ENST00000302692.7","ENST00000294454.6","ENST00000565488.6",
		   "ENST00000359511.5","ENST00000422440.7","ENST00000650617.1",
		   "ENST00000319017.5","ENST00000354883.11","ENST00000324194.12",
		   "ENST00000281154.6","ENST00000262999.4","ENST00000281456.11",
		   "ENST00000355943.8","ENST00000681962.1","ENST00000239451.7",
		   "ENST00000373627.10","ENST00000371347.10","ENST00000341119.10",
		   "ENST00000265631.10","ENST00000519973.6","ENST00000297578.9",
		   "ENST00000242275.7","ENST00000373069.10","ENST00000609923.6",
		   "ENST00000370495.6","ENST00000302503.8","ENST00000398802.6",
		   "ENST00000663595.2","ENST00000314032.9","ENST00000628067.3",
		   "ENST00000552981.6","ENST00000338625.9","ENST00000519676.6",
		   "ENST00000331299.6","ENST00000359232.8","ENST00000361529.5",
		   "ENST00000225665.12","ENST00000577745.2","ENST00000377095.10",
		   "ENST00000416858.7","ENST00000350690.10","ENST00000269205.7",
		   "ENST00000321510.7","ENST00000301454.9","ENST00000318596.8",
		   "ENST00000327451.11","ENST00000215882.10","ENST00000435456.7",
		   "ENST00000381401.11","ENST00000594199.3","ENST00000217909.8",
		   "ENST00000317881.9","ENST00000545805.6")

# Dataframe of just SLC25 read counts
dat2 <- subset(dat, dat$transcript_IDs %in% SLC25)
SLC25_dat <- subset(dat, dat$transcript_IDs %in% SLC25)

# Vector of the number of reads across transcripts originating from SLC25A4
reads <- dat2[,2]

# How many SLC25 transcripts were present in the Salmon index?
nrow(dat2) # 37

# Names of the SLC25 transcripts present in the index
yaxis_labels <- c("SLC25A38", "SLC25A51", "SLC25A50", "SLC25A16", "SLC25A43",
				  "SLC25A5", "SLC25A28", "SLC25A39", "SLC25A37", "SLC25A34",
				  "SLC25A29", "SLC25A32", "SLC25A1", "SLC25A20", "SLC25A4",
				  "SLC25A53", "SLC25A24", "SLC25A12", "SLC25A30", "SLC25A23",
				  "SLC25A31", "SLC25A40", "SLC25A19", "SLC25A35", "SLC25A33",
				  "SLC25A18", "SLC25A9", "SLC25A27", "SLC25A11", "SLC25A14",
				  "SLC25A22", "SLC25A15", "SLC25A7", "SLC25A46", "SLC25A10",
				  "SLC25A6", "SLC25A17")

# Bar plot of just the SLC25 transcripts 
pdf("output/SLC25_barplot.pdf")
bar <- barplot(reads,
			   main="Distribution of reads across the 37 SLC25 transcripts present",
               xlab="transcripts", ylim=c(0,50000), xaxt="n")
	   text(x=bar, y=-1600, yaxis_labels, cex=0.5, srt=45, xpd=TRUE)
dev.off()

# Dataframe of every other transcript read counts
'%!in%' <- Negate('%in%')
dat3 <- subset(dat, dat$transcript_IDs %!in% SLC25)

# Vector of the number of reads across all non-SLC25 transcripts
reads2 <- dat3[,2]

# How many transcrpts are there minus the SLC25 family?
nrow(dat3) # 177,419

# Bar plot of the distribution of reads across all other transcripts 
pdf("output/other_barplot.pdf")
bar2 <- barplot(reads2, main="Distribution of reads across 177,419 non-SLC25 transcripts",
               xlab="transcripts", ylim=c(0,50000))#, xaxt="n")
#	    text(x=bar2, y=-48, dat$transcript_IDs, cex=0.7, srt=45, xpd=TRUE)
dev.off()

# The number of reads simulated per transcript
# Instantiate vector of 0s equal to the number of SLC25 family members present
# in the Salmon index plus an index for the ambiguously mapped reads
# and the reads that did not map
expected <- rep(100, nrow(dat2))
expected[38] <- 0
expected[39] <- 0

# Which row has the read counts for SLC25A4 in the dataframe of only SLC25 members?
#which(dat2$transcript_IDs=="ENST00000281456.11") # 15 in dat2,  71073 in dat

# Read in the ambiguous mapping read count file generated by Salmon
ambig <- read.csv("mapped_reads/combined.fasta/aux_info/ambig_info.tsv", sep="\t")
nrow(ambig)==nrow(dat) # Are the number of transcript indices equal?; TRUE
multimap <- sum(ambig[,2])

# How many reads were ambiguously mapped in the SLC25 family?
slc25_ambig <- subset(ambig, rownames(ambig) %in% rownames(dat2)) 

# Add column with the transcript IDs and gene names
slc25_ambig$transcript_IDs <- dat2$transcript_IDs

# Write to csv
write.csv(slc25_ambig, "/scratch/mjpete11/salmon/output/ambiguous_slc25_map.csv")

# Reads that mapped incorrectly (e.g. not to SLC25A4)
wrong_map <- sum(dat3[,2])

# Vector of reads that mapped to SLC25 family members
slc25_map <- dat2[,2]

# Make a vector of the number of reads that mapped to the SLC25 transcripts, 
# the number that mapped elsewhere and the number that did not map 
experimental <- c(slc25_map, wrong_map, multimap) 

# Confusion matrix
tab2 <- table(experimental, expected)
tab2

tab <- table(factor(experimental, levels=seq(0,100,by=1)), 
			 factor(expected, levels=seq(0,100,by=1)))
tab[1:5,1:100]

# Convert to proportions
tab <- round(prop.table(tab), digits=2) 

# Change column and row names to the transcript IDs
rownames(tab) <- colnames(dat)[2:length(dat)]
colnames(tab) <- colnames(dat)[2:length(dat)]

# Write to csv
write.csv(tab, "/scratch/mjpete11/salmon/plots/confusion_matrix.csv")
